File: cdm-sources/src/cdm/margin/schedule/functions/IsIRSwaptionStraddle.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.product.template.EconomicTerms import EconomicTerms

__all__ = ['IsIRSwaptionStraddle']


@replaceable
def IsIRSwaptionStraddle(economicTerms: EconomicTerms) -> bool:
    """
    Identifies a product as an IR swaption straddle.
    
    Parameters 
    ----------
    economicTerms : EconomicTerms
    
    Returns
    -------
    is_Product : boolean
    
    """
    self = inspect.currentframe()
    
    
    is_Product =  (Qualify_InterestRate_Option_Swaption(rosetta_resolve_attr(self, "economicTerms")) and all_elements(rosetta_resolve_attr(get_only_element(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "economicTerms"), "payout"), "OptionPayout")), "optionType"), "=", rosetta_resolve_attr(OptionTypeEnum, "STRADDLE")))
    
    
    return is_Product

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)

File: cdm-sources/cdm/margin/schedule/functions/IsIRSwaptionStraddle.java
package cdm.margin.schedule.functions;

import cdm.product.qualification.functions.Qualify_InterestRate_Option_Swaption;
import cdm.product.template.EconomicTerms;
import cdm.product.template.OptionPayout;
import cdm.product.template.OptionTypeEnum;
import cdm.product.template.Payout;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.expression.CardinalityOperator;
import com.rosetta.model.lib.expression.ComparisonResult;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperS;
import javax.inject.Inject;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(IsIRSwaptionStraddle.IsIRSwaptionStraddleDefault.class)
public abstract class IsIRSwaptionStraddle implements RosettaFunction {
	
	// RosettaFunction dependencies
	//
	@Inject protected Qualify_InterestRate_Option_Swaption qualify_InterestRate_Option_Swaption;

	/**
	* @param economicTerms 
	* @return is_Product 
	*/
	public Boolean evaluate(EconomicTerms economicTerms) {
		Boolean is_Product = doEvaluate(economicTerms);
		
		return is_Product;
	}

	protected abstract Boolean doEvaluate(EconomicTerms economicTerms);

	public static class IsIRSwaptionStraddleDefault extends IsIRSwaptionStraddle {
		@Override
		protected Boolean doEvaluate(EconomicTerms economicTerms) {
			Boolean is_Product = null;
			return assignOutput(is_Product, economicTerms);
		}
		
		protected Boolean assignOutput(Boolean is_Product, EconomicTerms economicTerms) {
			is_Product = ComparisonResult.of(MapperS.of(qualify_InterestRate_Option_Swaption.evaluate(economicTerms))).and(areEqual(MapperS.of(MapperS.of(economicTerms).<Payout>mapC("getPayout", _economicTerms -> _economicTerms.getPayout()).<OptionPayout>map("getOptionPayout", payout -> payout.getOptionPayout()).get()).<OptionTypeEnum>map("getOptionType", optionPayout -> optionPayout.getOptionType()), MapperS.of(OptionTypeEnum.STRADDLE), CardinalityOperator.All)).get();
			
			return is_Product;
		}
	}
}
